cfscrape/__init__.py

```diff
@@ -94,48 +113,92 @@ def solve_cf_challenge(self, resp, **original_kwargs):
 
     def solve_challenge(self, body, domain):
         try:
-            js = re.search(r"setTimeout\(function\(\){\s+(var "
-                        "s,t,o,p,b,r,e,a,k,i,n,g,f.+?\r?\n[\s\S]+?a\.value =.+?)\r?\n", body).group(1)
+            challenge, ms = re.search(
+                r"setTimeout\(function\(\){\s*(var "
+                r"s,t,o,p,b,r,e,a,k,i,n,g,f.+?\r?\n[\s\S]+?a\.value\s*=.+?)\r?\n"
+                "(?:[^{<>]*},\s*(\d{4,}))?", body).groups()
+
+            # The challenge requires `document.getElementById` to get this content.
+            # Future proofing would require escaping newlines and double quotes
+            innerHTML = re.search(r"<div(?: [^<>]*)? id=\"cf-dn.*?\">([^<>]*)", body)
+            innerHTML = innerHTML.group(1) if innerHTML else ''
+
+            # Prefix the challenge with a fake document object.
+            # Interpolate the domain, div contents, and JS challenge.
+            # The `a.value` to be returned is tacked onto the end.
+            challenge = """
+                var document = {
+                    createElement: function () {
+                      return { firstChild: { href: "http://%s/" } }
+                    },
+                    getElementById: function () {
+                      return {"innerHTML": "%s"};
+                    }
+                  };
+
+                %s; a.value
+            """ % (domain, innerHTML, challenge)
+            # Encode the challenge for security while preserving quotes and spacing.
+            challenge = b64encode(challenge.encode("utf-8")).decode("ascii")
+            # Use the provided delay, parsed delay, or default to 8 secs
+            delay = self.delay or (float(ms) / float(1000) if ms else 8)
         except Exception:
             raise ValueError("Unable to identify Cloudflare IUAM Javascript on website. %s" % BUG_REPORT)
 
-        js = re.sub(r"a\.value = (.+ \+ t\.length(\).toFixed\(10\))?).+", r"\1", js)
-        js = re.sub(r"\s{3,}[a-z](?: = |\.).+", "", js).replace("t.length", str(len(domain)))
-
-        # Strip characters that could be used to exit the string context
-        # These characters are not currently used in Cloudflare's arithmetic snippet
-        js = re.sub(r"[\n\\']", "", js)
-
-        if "toFixed" not in js:
-            raise ValueError("Error parsing Cloudflare IUAM Javascript challenge. %s" % BUG_REPORT)
-
         # Use vm.runInNewContext to safely evaluate code
         # The sandboxed code cannot use the Node.js standard library
-        js = "console.log(require('vm').runInNewContext('%s', Object.create(null), {timeout: 5000}));" % js
+        js = """\
+            var atob = Object.setPrototypeOf(function (str) {\
+                try {\
+                    return Buffer.from("" + str, "base64").toString("binary");\
+                } catch (e) {}\
+            }, null);\
+            var challenge = atob("%s");\
+            var context = Object.setPrototypeOf({ atob: atob }, null);\
```

Comment:
It's probably better to continue the discussion here and/or request changes for other issues.